Improve AcroForm/XFA form type detection

timvandermeij opened the pull request:
The commit messages contain more information about the individual changes.

Fixes #12217.
Replaces #12254.

For completeness, here is the list of test PDF files with the form type detection results from before/after this patch (note that we only fallback if _only_ XFA is available):

| File  | Before | After | Notes |
| ------ | -------- | ------- | ------- |
| http://www.aloaha.com/wp-content/uploads/2016/07/SampleForm-1.pdf  | AcroForm, XFA, fallback | AcroForm, XFA, no fallback|
| http://www.cic.gc.ca/english/pdf/kits/forms/IMM5257E.PDF  | AcroForm, XFA, fallback | XFA, fallback | `SigFlags` bit set and nested signature in `Fields`.
| https://web.archive.org/web/20121105185256if_/http://www.northeastern.edu/hrm/pdfs/resources/benefits/MBTA-pretax-form-July2012.pdf | AcroForm, XFA, fallback | XFA, fallback | Empty `Fields`.
| https://web.archive.org/web/20110918100215/http://www.irs.gov/pub/irs-pdf/f1040.pdf | AcroForm, no fallback | AcroForm, no fallback | `SigFlags` available, but first bit not set.
| https://github.com/mozilla/pdf.js/files/762326/212241.6.pdf | AcroForm, no fallback | AcroForm, no fallback | `SigFlags` bit set and multiple signatures in `Fields`.

timvandermeij committed:
Move the version logic from the document to the catalog

The `Version` entry is part of the catalog, not of the document, so its
logic should be placed there instead. The document should look in the
catalog to fetch it, and not have knowledge of `catDict`, which is a
member internal to the catalog.

Moreover, make the version member private on the document instance. It's
only used internally and was also never intended to be public. For users
it's exposed by the `getMetadata` API endpoint as `PDFFormatVersion`.

Finally, clarify how the version from the header and the version from
the catalog are treated using a comment.

timvandermeij committed:
Move the collection logic from the document to the catalog

The `Collection` entry is part of the catalog, not of the document, so
its logic should be placed there instead. The document should look in the
catalog to fetch it, and not have knowledge of `catDict`, which is a
member internal to the catalog.

Moreover, remove the collection member from the document instance. It's
only used internally and was also never intended to be public. For users
it's exposed by the `getMetadata` API endpoint as `IsCollectionPresent`.
Moving this out of the `parse` function makes sure that the getter is
only executed if the document information is actually requested
(potentially making initial parsing a tiny bit faster).

timvandermeij committed:
Move the AcroForm logic from the document to the catalog

The `AcroForm` entry is part of the catalog, not of the document, so its
logic should be placed there instead. The document should look in the
catalog to fetch it, and not have knowledge of `catDict`, which is a
member internal to the catalog.

Moreover, make the AcroForm member private on the document instance. It's
only used internally and was also never intended to be public. For users
it's exposed by the `getMetadata` API endpoint as `IsAcroFormPresent`.
Only a boolean is exposed, so we now also only store the boolean on the
document instance.

Finally, the annotation code needs access to the full AcroForm
dictionary, so it's updated to fetch the data from the catalog instead
of the document that now only holds the boolean.

timvandermeij committed:
Mark the `catDict` member as private in the `Catalog` class

Not only is `catDict` never accessed anymore outside of this file, it
should also never happen since it's internal to the catalog. If data
from it is needed elsewhere, the catalog should provide a getter for it
that can do basic data integrity checks and abstract away any
unnecessary details.

timvandermeij committed:
Redo the form type detection logic and include unit tests

Good form type detection is important to get reliable telemetry and to
only show the fallback bar if a form cannot be filled out by the user.

PDF.js only supports AcroForm data, so XFA data is explicitly unsupported
(tracked in issue #2373). However, the previous form type detection
couldn't separate AcroForm and XFA well enough, causing form type
telemetry to be incorrect sometimes and the fallback bar to be shown for
forms that could in fact be filled out by the user.

The solution in this commit is found by studying the specification and
the form documents that are available to us. In a nutshell the rules are:

- There is XFA data if the `XFA` entry is a non-empty array or stream.
- There is AcroForm data if the `Fields` entry is a non-empty array and
  it doesn't consist of only document signatures.

The document signatures part was not handled in the old code, causing a
document with only XFA data to also be marked as having AcroForm data.
Moreover, the old code didn't check all the data types.

Now that AcroForm and XFA can be distinguished, the viewer is configured
to only show the fallback bar for documents that only have XFA data. If
a document also has AcroForm data, the viewer can use that to render the
form. We have not found documents where the XFA data was necessary in
that case.

Finally, we include unit tests to ensure that all cases are covered and
move the form type detection out of the `parse` function so that it's
only executed if the document information is actually requested
(potentially making initial parsing a tiny bit faster).

timvandermeij commented:
/botio test

Snuffleupagus commented on src/core/document.js:
```diff
@@ -715,6 +697,61 @@ class PDFDocument {
+    return fields.every(field => {
+      field = this.xref.fetchIfRef(field);
+      if (field.has("Kids")) {
+        return this._hasOnlyDocumentSignatures(field.get("Kids"));
```
Given that this is a *recursive* function call, you really must protect against infinite loops in this method!

One option could be to track all seen `Kids` nodes and break if anyone is ever repeated, however that relies on the assumption that they're always `Ref`s which may not be guaranteed in the general case.
The easiest solution is probably to just keep track of the recursion-depth, and break when it reaches some reasonable limit (e.g. `10`?).

Snuffleupagus commented on src/core/document.js:
```diff
@@ -715,6 +697,61 @@ class PDFDocument {
+      if (field.has("Kids")) {
+        return this._hasOnlyDocumentSignatures(field.get("Kids"));
+      }
+      const fieldType = field.get("FT");
+      const isSignature = isName(fieldType) && fieldType.name === "Sig";
```
This can be actually be simplified to the e.g. following:
```suggestion
      const isSignature = isName(/* field type = */ field.get("FT"), "Sig");
```

Snuffleupagus commented on src/core/document.js:
```diff
@@ -744,7 +742,7 @@ class PDFDocument {
       IsAcroFormPresent: !!this.acroForm,
       IsXFAPresent: !!this.xfa,
-      IsCollectionPresent: !!this.collection,
+      IsCollectionPresent: this._hasCollection,
```
With the changes to the `Catalog.collection` getter, this can be simplified to *only* the following and `this._hasCollection` removed since it's unused (in later commits, you're also removing the temporarily added `this._hasAcroForm` boolean):
```suggestion
      IsCollectionPresent: !!this.catalog.collection,
```

Snuffleupagus commented on src/core/document.js:
```diff
@@ -715,6 +697,61 @@ class PDFDocument {
+        const fields = acroForm.get("Fields");
+        const hasFields = Array.isArray(fields) && fields.length > 0;
+        const sigFlags = acroForm.get("SigFlags");
+        const hasDocumentSignatures = !!(sigFlags & 0x1);
+        let hasOnlyDocumentSignatures = false;
+        if (hasDocumentSignatures) {
+          hasOnlyDocumentSignatures = this._hasOnlyDocumentSignatures(fields);
+        }
```
Nit: You can probably collapse this a little bit, since there's intermediate state that's not really being used here.
```suggestion
        const hasOnlyDocumentSignatures =
          !!(sigFlags & 0x1) && this._hasOnlyDocumentSignatures(fields);
          
```

Snuffleupagus reviewed (changes requested):
This definitely looks like a big step in the right direction, given how complicated the AcroForm/XFA situation apparently is; however there's a couple of things I'd suggest changing.

I've not completely reviewed all of this code yet, notably the new unit-tests, but figured I'd wait with that until the patches are updated.

Snuffleupagus commented on src/core/document.js:
```diff
@@ -714,6 +697,61 @@ class PDFDocument {
+
+  get formInfo() {
+    const formInfo = { hasAcroForm: false, hasXfa: false };
+    try {
+      const acroForm = this.catalog.acroForm;
+      if (acroForm) {
```
Also, with the error handling added to `Catalog.acroForm` you can swap these lines around a little bit:
```suggestion
    const acroForm = this.catalog.acroForm;
    if (acroForm) {
      try {
```

timvandermeij commented:
Thank you for the review! I have addressed all comments in the new commit series.

Snuffleupagus commented on src/core/document.js:
```diff
@@ -574,40 +575,12 @@ class PDFDocument {
   parse(recoveryMode) {
     this.setup(recoveryMode);
```
Now that you have *significantly* simplified the `parse` method, and thank you for doing that since it always felt a bit unwieldy, I believe that you can simply inline `PDFDocument.setup` here given that it only contains two lines of code:
https://github.com/mozilla/pdf.js/blob/34cec29288212a6ca5bb9d703d6c5102c47e635d/src/core/document.js#L706-L709

Snuffleupagus reviewed (approved):
I've added three more comments, after which I believe that this is good to go :-) 

This code now looks much nicer, thank you!

Snuffleupagus commented on src/core/document.js:
```diff
@@ -714,6 +687,67 @@ class PDFDocument {
+      // The document contains XFA data if the `XFA` entry is a non-empty
+      // array or stream.
+      const xfa = acroForm.get("XFA");
+      const hasXfa = (Array.isArray(xfa) || isStream(xfa)) && xfa.length > 0;
```
One final thing: When the `XFA`-entry is a Stream, can it ever be a something that's (indirectly) a `DecodeStream`?
In that case, there's no concept of exact Stream-length and consequently no `length` getter. Hence I believe that to prevent *any* future issues here, you want the following instead which is guaranteed to work for *all* Stream-implementations:
```suggestion
      const hasXfa =
        (Array.isArray(xfa) && xfa.length > 0) ||
        (isStream(xfa) && !xfa.isEmpty);
```

timvandermeij replied:
The specification is not really clear on that, but since it wouldn't be the first time we see unexpected things in PDF files, I'll indeed handle this case. Moreover, the `isEmpty` getter is nicer since it works in all cases; I didn't know we had that.

timvandermeij commented on src/core/document.js:
```diff
@@ -697,6 +687,67 @@ class PDFDocument {
+      const isSignature = isName(field.get("FT"), "Sig");
+      const rectangle = field.get("Rect");
+      const isInvisible =
+        Array.isArray(rectangle) && rectangle.every(value => value === 0);
```
I found this while looking at the various files, but I looked up some more documentation on this and found https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/acrobat_digital_signature_appearances_v9.pdf#page=7&zoom=auto,-216,358 that mentions that "for invisible signatures, the annotation has a rectangle array with values [ 0 0 0 0 ]", so the current code should be correct as it matches the specification and what I found in the files.

timvandermeij committed:
Inline the `setup` method in the `parse` method in `src/core/document.js`

Now that the `parse` method is simplified we can inline the `setup`
method in the `parse` method since it's only two lines of code. This
avoids some indirection.

timvandermeij commented:
/botio unittest

timvandermeij commented:
/botio makeref
